### PR TITLE
material fishing rods and material fish.

### PR DESCRIPTION
**Pull request by Ghommie**
## About The Pull Request
The reason why I refactored material effects code is here.

In this PR I add the possibility of printing fishing rods from different materials at your nearest autolathe, though it doesn't stop there. Each material has different values for (so far) fishing difficulty, casting range, experience gained and foremost the chance of catching fish made of the same material as the fishing rod used to catch it. The material the fish is made from can either increase or decrease the weight of the fish.

In a many cases, material rods can also have other small effects on fishing, like removing the chances of duds even without a bait, or attracting shiny-loving fish, or passing some other fish trait checks. In a few cases, these fishing rods can have bigger, more impactful effects on fishing. That said, here's a list of more or less the effects (skipping very minor ones and those inherited from being items) and modifiers across materials so far, a few things will be added later:

<details>
 <summary>Huge-ass list (out-fucking-dated now that I've added 5 more variables 😢)</summary>

- Iron: +30% fish weight, 8% chance of material fish
- Glass: +20% fish weight, +5 difficulty, +20% experience, 8% chance of material fish
- Silver: +35% fish weight, -5 difficulty, - 15% experience, 15.5% chance of material fish
- Gold: +50% fish weight, -10 difficulty, -25% experience, 20.5% material fish, +1 casting range
- Diamond: +40% fish weight, -13 difficulty, -30% experience, 23% material fish, -1 casting range
- Uranium: +100% fish weight, 8% material fish
- Plasma: +30% fish weight, 8% material fish
- Bluespace: +30% fish weight, -5 difficulty, -15% experience, 23% material fish, +5 range
- - 33% chance of selecting a reward from (almost) any fishing source instead of the current one.
- Bananium +20 difficulty, +60% experience, 38% material fish, +3 casting range
- - 20% chance of fishing either a clownfish, lubefish, donkfish, soulfish or skin crab instead
- Titanium: +20% fish weight, -5 difficulty, 8% material fish, +1 range
- Runite: +50% fish weight, -18 difficulty, +220% experience, 38% material fish, +1 range
- Plastic: -20% fish weight -5 difficulty, +20% experience, 8% material fish, +2 range
- Wood: -50% fish weight, +8 difficulty, +30% experience, 13% material fish, -1 range
- Adamantine: +60% fish weight, -23 difficulty, -40% experience, 33% material fish, +1 range
- Mythril: +40% fish weight, -25 difficulty, -50% experience, 43% material fish, +2 range
- Hot Ice: -10% fish weight, -10 diffiulty, -10% experience, 18% material fish, +1 range
- Metal Hydrogen: -40% weight, -15 difficulty, -20% experience, 23% material, +4 range
- Sand: +20% weight, +30 difficulty, -80% experience, 8% material, -2 range
- Sandstone: +20% weight, +25 difficulty, -70% experience, 8% material, -2 range
- Snow: -20% weight, +25 difficulty, -70% experience, 8% material, -2 range
- Runed Metal: +50% weight, -12 diff, -10% experience, 38% material
- Bronze, +40% weight, 13% material
- Paper: -60% weight, +40 diff, -90% experience, 8% material, -2 range
- - Has a 20% chance of fishing up an aggressive stickman
- Cardboard: Same as above, but without the stickmen
- Bone: +5% weight, +15 diff, -15% experience, 8% material, -2 range
- - 20% chance of fishing up either a unmarine bonemass, a unmarine mastodon, a hostile skellie or rarely a single-use spectral instrument that can turn you or someone else into a spooky scary skeleton.
- Bamboo: -50% weight, -4 difficulty, +30% experience, 13% material, -1 range
- Zeukerite: 20% weight, -16 difficulty, -10% experience, 28% material
- Plasteel: +75% weihgt, +5 difficulty, +10% experience, 8% material
- Plastitanium: +10% weight, -10 difficulty, -5% experience, 8% material, +1 range
- Plasmaglass: +20% weight, +5 difficulty, +30% experience, 8% material
- Titanium Glass: +25% weight, -5 difficulty, +25% experience 8% material
- Plastitanium Glass: +20% weight, +50% experience, 8% material
- Alien Alloy: +140% weight, -25 difficulty, 33% material, -40% experience, +2 range
- Hauntium (good luck using it): +40% weight, -30 difficulty, +50% experience, 38% material, +2 range
- Meat: +13 difficulty, -20% experience, 28% material, -2 range
- Pizza: -10% weight, +13 difficulty, -20% experience, 28% material, -2 range

</details>

## Why It's Good For The Game
I think it's nice to have two features interact with each other (fishing and materials), and it adds a layer of interesting uses to some materials.

## Changelog

:cl:
add: Autolathes can now print fishing rods of different materials, which can increase or decrease fishing difficulty, casting range, and experience gained, and can have other interesting effects.
/:cl:

### CODE REVIEW

**Comment by Eronymun** [comment]
This is really awesome. I love some of the unique effects too, would love to see more of them. A few effects id like to see listed below.
Alien alloy, as is tradition with alien tools, should be really fast, either with catching or having your bobber be extra fast.
You should be able to feed milk to bone material fish to heal them, for consistency.
Sand and sandstone fish should be heat resistant and maybe turn into glass fish when hot enough...
Zaukerite rods should put a little poison in every fish they catch, maybe these fish should also generate tons of power if fed to the sm, for those Atmos players.
This one is insane, but shooting bamboo fish out of the bamboo blow gun.
I don't know which one you would put this on but being able to throw out two lines at once on any fishing pool would be cool.

Just some random ideasguysing.

**Comment by Loafin34** [comment]
could plasma fishing rods catch lava-based fish?

**Comment by Ghommie** [comment]
> could plasma fishing rods catch lava-based fish?

I guess I can do that.

**Comment by CCC23** [comment]
now all I have to do is figure out how to get mythril fishing rods and I can get another mythril fishing rod

File: code/datums/materials/_material.dm [hunk 1]
```diff
@@ -100,7 +124,23 @@ Simple datum which is instanced once per type and is used for every object of sa
...
+	RegisterSignal(source, COMSIG_ITEM_FISHING_ROD_SLOTTED, PROC_REF(on_fishing_rod_slotted))
+	RegisterSignal(source, COMSIG_ITEM_FISHING_ROD_UNSLOTTED, PROC_REF(on_fishing_rod_unslotted))
+	if(istype(source, /obj/item/fishing_rod))
+		ADD_TRAIT(source, TRAIT_ROD_ATTRACT_SHINY_LOVERS, REF(src))
+	if(istype(source.loc, /obj/item/fishing_rod))
+		on_fishing_rod_slotted(source, source.loc)
+
+/datum/material/proc/on_fishing_rod_slotted(datum/source, obj/item/fishing_rod/rod, slot)
+	SIGNAL_HANDLER
+	ADD_TRAIT(rod, TRAIT_ROD_ATTRACT_SHINY_LOVERS, REF(source))
```
Comment (SmArtKar) [inline code/datums/materials/_material.dm:139]: Probably shouldn't affect all mats, wood or paper aren't particularly shiny

File: code/datums/materials/basemats.dm [hunk 2]
```diff
@@ -258,6 +315,43 @@ Unless you know what you're doing, only use the first three numbers. They're in
...
+	))
+	var/datum/fish_source/new_source
+	var/list/elegible_fish_sources = GLOB.preset_fish_sources.Copy()
+	for(var/source_type in elegible_fish_sources)
+		var/datum/fish_source/source = elegible_fish_sources[source_type]
+		if(source.fish_source_flags & FISH_SOURCE_FLAG_NO_BLUESPACE_ROD)
+			elegible_fish_sources -= source_type
+	new_source = elegible_fish_sources[pick(elegible_fish_sources)]
+	new_source.RegisterSignal(challenge, COMSIG_FISHING_CHALLENGE_ROLL_REWARD, TYPE_PROC_REF(/datum/fish_source, roll_reward_minigame))
+	new_source.RegisterSignal(challenge, COMSIG_FISHING_CHALLENGE_GET_DIFFICULTY, TYPE_PROC_REF(/datum/fish_source, calculate_difficulty_minigame))
```
Comment (SmArtKar) [inline code/datums/materials/basemats.dm:349]: Material shouldn't be handling source's comsigs, please do something about this, maybe add some way to override the source instead?

File: code/game/objects/items.dm [hunk 3]
```diff
@@ -1824,6 +1824,16 @@
...
 
+///Called when an item is added to a slot of a fishing rod.
+/obj/item/proc/on_fishing_rod_slotted(obj/item/fishing_rod/rod, slot)
+	SHOULD_CALL_PARENT(TRUE)
+	SEND_SIGNAL(src, COMSIG_ITEM_FISHING_ROD_SLOTTED, rod, slot)
+
+///Called when an item is removed from a slot of a fishing rod
+/obj/item/proc/on_fishing_rod_unslotted(obj/item/fishing_rod/rod, slot)
+	SHOULD_CALL_PARENT(TRUE)
+	SEND_SIGNAL(src, COMSIG_ITEM_FISHING_ROD_UNSLOTTED, rod, slot)
```
Comment (SmArtKar) [inline code/game/objects/items.dm:1835]: This has no place on /obj/item, put these on smth like /obj/item/fishing and repath (and make an UpdatePaths script) for lured and lines and what not uses it

File: code/modules/fishing/fish/fish_traits.dm [hunk 4]
```diff
@@ -530,6 +551,11 @@ GLOBAL_LIST_INIT(spontaneous_fish_traits, populate_spontaneous_fish_traits())
 	added_difficulty = 5
 	reagents_to_add = list(/datum/reagent/lube = 1.2)
 
+/datum/fish_trait/lubed/catch_weight_mod(obj/item/fishing_rod/rod, mob/fisherman, atom/location, obj/item/fish/fish_type)
+	. = ..()
+	if(GetComponent(rod, /datum/component/slippery)) //At least you get x5 chance of catching a lubefish with a bananium rod.
```
Comment (SmArtKar) [inline code/modules/fishing/fish/fish_traits.dm:556]: veeery bad practice

**Review by SmArtKar — CHANGES_REQUESTED** [review]

File: code/datums/materials/_material.dm [hunk 5]
```diff
@@ -78,6 +78,30 @@ Simple datum which is instanced once per type and is used for every object of sa
...
+	/// The multiplier to the completion gain of the fishing rod made of this material
+	var/fishing_completion_speed = 1
+	/// The multiplier of the bait/bobber speed of the fishing challenge for fishing rods made of this material
+	var/fishing_bait_speed_mult = 1
+	/// The multiplier of the deceleration/friction for fishing rods made of this material
+	var/fishing_deceleration_mult = 1
+	/// The multiplier of the bounciness of the bait/bobber upon hitting the edges of the minigame area
+	var/fishing_bounciness_mult = 1
+	/// The multiplier of negative velocity that pulls the bait/bobber of a fishing rod down when not holding the click
+	var/fishing_gravity_mult = 1
```
Comment (SmArtKar) [inline code/datums/materials/_material.dm:103]: Reposting from discord, this is half the length of our current vars if not more - you need to cut down on these, perhaps replace some of them with existing material properties

File: code/datums/materials/_material.dm (hunk 1 above)
Comment (Ghommie) [reply code/datums/materials/_material.dm:139]: This proc is only registered if the material has a certain var set to TRUE.

File: code/game/objects/items.dm (hunk 3 above)
Comment (Ghommie) [reply code/game/objects/items.dm:1835]: I can move these and some of the other procs in a different file, perhaps in the fishing module, but I don't agree with this, kind of. The bait slot can take any kind of item as long as they have the bait trait, even if it seldom has any additional effect.

**Comment by Wallemations** [comment]
In case new materials are added, do all materials not covered by this PR default to normal stats?

**Comment by Ghommie** [comment]
Only if the author of the new material doesn't set any variable, which is on him.

File: code/modules/fishing/fish/_fish.dm [hunk 6]
```diff
@@ -440,24 +448,36 @@
 		. += span_warning("It's been bitten by someone.")
 
 ///Randomizes weight and size.
-/obj/item/fish/proc/randomize_size_and_weight(base_size = average_size, base_weight = average_weight, deviation = weight_size_deviation)
+/obj/item/fish/proc/randomize_size_and_weight(base_size = average_size, base_weight = average_weight, deviation = weight_size_deviation, update = TRUE)
```
Comment (ArcaneMusic) [inline code/modules/fishing/fish/_fish.dm:451]: I would love a slightly more accurate comment here

File: code/modules/reagents/chemistry/reagents/toxin_reagents.dm [hunk 7]
```diff
@@ -1230,6 +1230,11 @@
 			to_chat(affected_mob, span_warning("A phantom limb hurts!"))
 			affected_mob.say("Why are we still here, just to suffer?", forced = /datum/reagent/toxin/bonehurtingjuice)
 
+/datum/reagent/milk/used_on_fish(obj/item/fish/fish)
+	if(HAS_TRAIT(fish, TRAIT_FISH_MADE_OF_BONE))
+		fish.adjust_health(fish.health - 30)
+		return TRUE
```
Comment (ArcaneMusic) [inline code/modules/reagents/chemistry/reagents/toxin_reagents.dm:1236]: Looks like a duplicate of `code/modules/reagents/chemistry/reagents/drinks/drink_reagents.dm` line 229, pick one

**Review by ArcaneMusic — CHANGES_REQUESTED** [review]
I found a handful of relatively minor things, but all in all it's quite late and I have no clue how 90% of fishing works at present so this has only exacerbated the issue

File: code/modules/reagents/chemistry/reagents/toxin_reagents.dm (hunk 7 above)
Comment (Ghommie) [reply code/modules/reagents/chemistry/reagents/toxin_reagents.dm:1236]: Ah, this was meant for the bone hurting juice. Good effing catch!

**Review by ArcaneMusic — APPROVED** [review]
Nothing I saw looked like it could do any damage, but considering the number of numbers flying around WRT the materials on each specific fishing rod, I can foresee either number tweaking in your horizon on a number of materials, or potentially trimming down the fishing-related material qualities. Food for thought.

**Pull request merged**